aesara/graph/opt.py

```diff
@@ -1697,7 +1693,16 @@ def transform(self, fgraph, node, get_nodes=True):
         If it does, it constructs ``out_pattern`` and performs the replacement.
 
         """
-        from aesara.graph import unify
+        from etuples.core import ExpressionTuple
+        from unification import reify, unify
+
+        # TODO: We shouldn't need to iterate like this.
+        if not self.allow_multiple_clients and any(
+            len(fgraph.clients.get(v)) > 1
+            for v in vars_between(fgraph.inputs, node.outputs)
+            if v not in fgraph.inputs
+        ):
+            return False
```

Comment:
Won't this iteration restrict the graphs to have single clients till top level inputs of the graph ? and not the inputs to the pattern we're trying to detect ? or was the former the intention to begin with ?

Reply:
If I remember the current/old logic correctly, it would check `allow_multiple_clients` as it walked up the graph during unification; now, we're performing that check before attempting to unify the graphs.

It would be a bit more efficient to do this check iteratively instead of making two passes (that's part of the reason behind my TODO comment), but I believe it performs the same check in both cases.

Otherwise, it's not entirely clear what that option is supposed to mean at a high-level, but I'm assuming&mdash;when `False`&mdash;it means that a graph shouldn't be replaced if it contains terms with more than one client.  I don't know what it's set to `False` by default, though, because that seems oddly restrictive and likely to cause confusion.

Reply:
Yeah I think the default should be True, that's how most local optimizers work.